File: client/lib/game/core/command.dart
import 'dart:mirrors';
import 'package:client/game/core/abstract/game_object.dart';

class Command
{
  String targetID;
  String action;
  List args;

  Command(this.targetID, this.action, [this.args = const []]){}

  Command.from(Map serialization)
  {
    targetID = serialization['targetID'];
    action = serialization['action'];
    args = serialization['args'];
  }

  execute(GameObject target)
  {
    //Invoke a command on a given GameObject
    InstanceMirror mirror = reflect(target);
    mirror.invoke(MirrorSystem.getSymbol(action), args);
  }

  Map serialize()
  {
    return {'targetID': targetID, 'action': action, 'args': args};
  }
}

File: src/game/game_core/command.py
class Command(object):
    targetID = None
    action = None
    args = None

    def __init__(self, targetid, action, args=[]):
        if targetid and (isinstance(targetid, bytes) or isinstance(targetid, str)):
            self.targetID = targetid
        else:
            raise TypeError
        if action and isinstance(action, str):
            self.action = action
        else:
            raise TypeError
        if isinstance(args, list):
            self.args = args
        else:
            raise TypeError

    def execute(self, target):
        if target.ID == self.targetID:
            getattr(target, self.action)(*self.args)
        else:
            raise NameError

    def serialize(self):
        return [self.targetID, self.action, self.args]
